main/admin/init_cas_user_extra_field_from_ldap.php

```diff
@@ -0,0 +1,323 @@
+                ||
+                strtolower(trim($username)) === strtolower(trim($ldapCASUser))
+            ) {
+                if (array_key_exists($ldapUsername, $userNamesInUse)) {
+                    print "wrong username but '$ldapUsername' is already taken, skipping.\n";
+                    break;
+                } else {
+                    $problems[] = "wrong username";
+                    $wrongUserNames[$userId] = $ldapUsername;
+                    $userNamesInUse[$ldapUsername] = $userId;
+                }
+            } else {
+                die("LDAP search result does not match username; our filter is wrong: $filter\n");
+            }
+            if (array_key_exists($userId, $existingCasUserValues)) {
+                $currentValue = $existingCasUserValues[$userId];
+                if ($currentValue !== $ldapCASUser) {
+                    $problems[] = "wrong current CAS user code '$currentValue'";
+                    $wrongCASCodes[$userId] = $ldapCASUser;
+                }
+            } else {
+                $problems[] = "CAS user code missing in database";
+                $missingCASCodes[$userId] = $ldapCASUser;
+            }
+            $currentAuthSource = $user->getAuthSource();
+            if (CAS_AUTH_SOURCE === $currentAuthSource) {
+            } else {
+                $problems[] = "wrong auth source '$currentAuthSource'";
+                $wrongAuthSources[$userId] = true;
+            }
+            print (empty($problems) ? "ok\r" : (join(', ', $problems)."\n"));
+            break;
+        default:
+            print "more than one entries for username '$username' in the LDAP directory for user id=$userId, skipping.\n";
+    }
+    $checked ++;
+}
+print "\033[2K\r";
+
+
+// ask for confirmation and write changes to the database
+
+$fixUsernames = (
+    !empty($wrongUserNames)
+    &&
+    ('y' === readline(
+            "Fix wrong user names for ".count($wrongUserNames)." users ?"
+            . " (type 'y' to confirm) "
+        )
+    )
+);
+if ($fixUsernames) print "I will fix user names.\n";
+
+$fixMissingCASCodes = (
+    !empty($missingCASCodes)
+    &&
+    ('y' === readline(
+            "Fix missing CAS codes for ".count($missingCASCodes)." users ?"
+            . " (type 'y' to confirm) "
+        )
+    )
+);
+if ($fixMissingCASCodes) print "I will fix missing CAS codes.\n";
+
+$fixWrongCASCodes = (
+    !empty($wrongCASCodes)
+    &&
+    ('y' === readline(
+            "Fix wrong CAS codes for ".count($wrongCASCodes)." users ?"
+            . " (type 'y' to confirm) "
+        )
+    )
+);
+if ($fixWrongCASCodes) print "I will fix wrong CAS codes.\n";
+
+$fixWrongAuthSources = (
+    !empty($wrongAuthSources)
+    &&
+    ('y' === readline(
+            "Fix auth source for ".count($wrongAuthSources)." users ?"
+            . " (type 'y' to confirm) "
+        )
+    )
+);
+if ($fixWrongAuthSources) print "I will fix wrong authentication sources.\n";
```

Comment:
I think this script can remain with hardcoded English messages because it is not part of the user interface. It is supposed to be run only once and for all in a terminal and only in rare cases (LDAP to CAS migration).

Reply:
ok